src/block_cache/block_mgr.c

```diff
@@ -123,38 +90,41 @@ static int
 __bm_checkpoint(
   WT_BM *bm, WT_SESSION_IMPL *session, WT_ITEM *buf, WT_CKPT *ckptbase, bool data_checksum)
 {
-    WT_BLOCK *block, *tblock;
-    WT_CONNECTION_IMPL *conn;
+    WT_BLOCK *block;
+    u_int i;
     bool found;
 
-    conn = S2C(session);
     block = bm->block;
 
     WT_RET(__wt_block_checkpoint(session, block, buf, ckptbase, data_checksum));
 
+    if (!bm->is_multi_handle)
+        return (0);
     /*
-     * Close previous primary objects that are no longer being written, that is, ones where all
-     * in-flight writes have drained. We know all writes have drained when a subsequent checkpoint
-     * completes, and we know the metadata file is the last file to be checkpointed. After
-     * checkpointing the metadata file, review any previous primary objects, flushing writes and
-     * discarding the primary reference.
+     * For tiered tables, we need to fsync any previous active files to ensure the full checkpoint
+     * is persisted. We wait until now because there may have been in-progress writes to old files.
+     * But now we know those writes must have completed. Checkpoint ensures that all dirty pages of
+     * the tree have been written and eviction is disabled at this point, so no new data is getting
+     * written.
+     *
+     * We don't hold the handle table lock across fsync calls since those could be slow and that
+     * would block a concurrent thread opening a new block handle.
      */
-    if (strcmp(WT_METAFILE, block->name) != 0)
-        return (0);
     do {
         found = false;
-        __wt_spin_lock(session, &conn->block_lock);
-        TAILQ_FOREACH (tblock, &conn->blockqh, q)
-            if (tblock->close_on_checkpoint) {
-                tblock->close_on_checkpoint = false;
-                __wt_spin_unlock(session, &conn->block_lock);
+        __wt_readlock(session, &bm->handle_table_lock);
+        for (i = 0; i < bm->handle_table_next; ++i) {
+            block = bm->handle_table[i];
+            if (block->sync_on_checkpoint) {
                 found = true;
-                WT_RET(__wt_fsync(session, tblock->fh, true));
-                WT_RET(__bm_close_block(session, tblock));
+                __wt_readunlock(session, &bm->handle_table_lock);
+                WT_RET(__wt_fsync(session, block->fh, true));
+                block->sync_on_checkpoint = false;
                 break;
             }
+        }
     } while (found);
-    __wt_spin_unlock(session, &conn->block_lock);
+    __wt_readunlock(session, &bm->handle_table_lock);
```

Comment:
Good point! I've restructured the code along the lines you suggest and it is definitely clearer. 

Maybe this isn't exactly what you were thinking since the `__wt_readunlock` doesn't wind up immediately before the `while (found)`. We have to call `fsync` in there without the lock. 

Let me know, @ddanderson, if you had something different in mind.

Reply:
Yes, that's what I had in mind, I messed up the details.  Thanks!